[nativert] Free stale execution frames

georgiaphillips opened the pull request:
Summary:
This was implemented in SR due to caching of runtime instances building up and causing some memory usage spikes after some large amount traffic went through the model, and then once traffic went down, SR was still caching all the previous usage.

We need something similar on the Sigmoid side to make sure the static dispatch modules aren't hogging memory. Currently, all ExecutionFrame objects are being cached, and never freed if stale.

Test Plan:
Added extra execution frames in tmp commit D75257998 and ran local replayer test to confirm extra execution frames get cleaned up down to min size, which is set at 8

 {F1978532047}


Also tested by modifying load_net_predictor (modifications also in D75257998) to run benchmarkNumIterations twice - once with benchmarkNumThreads, and once with only one thread. Also set clearing interval at one second. Verified that execution frames get cleared when we drop down to one thread.

 {F1978558984}


```
buck2 test 'mode/dev-nosan' fbcode//sigmoid/inference/test_gpu:model_runner_test -- ModelRunnerTest.Basic_InterpreterCuda_Multithread_Cleanup --run-disabled --print-passing-details
```

Differential Revision: D75257992

zhxchen17 commented on torch/nativert/executor/ExecutorConfig.h:
```diff
@@ -16,6 +16,8 @@ struct ExecutorConfig {
+  int64_t minNumExecutionFrames = 1;
+  int64_t executionFramePoolCleanupIntervalSec = 600;
 };
 
```
looks good. title needs to be changed.

georgiaphillips committed:
[Sigmoid] Free stale execution frames (#154636)

Summary:

This was implemented in SR due to caching of runtime instances building up and causing some memory usage spikes after some large amount traffic went through the model, and then once traffic went down, SR was still caching all the previous usage.

We need something similar on the Sigmoid side to make sure the static dispatch modules aren't hogging memory. Currently, all ExecutionFrame objects are being cached, and never freed if stale.

Test Plan:
Added extra execution frames in tmp commit D75257998 and ran local replayer test to confirm extra execution frames get cleaned up down to min size, which is set at 8

 {F1978532047}


Also tested by modifying load_net_predictor (modifications also in D75257998) to run benchmarkNumIterations twice - once with benchmarkNumThreads, and once with only one thread. Also set clearing interval at one second. Verified that execution frames get cleared when we drop down to one thread.

 {F1978558984}


```
buck2 test 'mode/dev-nosan' fbcode//sigmoid/inference/test_gpu:model_runner_test -- ModelRunnerTest.Basic_InterpreterCuda_Multithread_Cleanup --run-disabled --print-passing-details
```
MODEL_ENTITY_ID=727362647
SNAPSHOT_ID=2

ICE run for CPU model w/o changes: MDS_georgiaphillips_1748545099_b84d
https://fburl.com/unidash/r1dlwc3e


ICE run w changes: MDS_georgiaphillips_1748541785_aac4
https://fburl.com/unidash/c0niwj7f

[Memory] A runMultiThenSingleThread Turnover: 8,973,376

Reviewed By: dolpm

Differential Revision: D75257992
